src/lib.rs

```diff
@@ -23,25 +23,394 @@ extern crate assert_matches;
+                },
+            )?);
+        }
+
+        // TODO: this is too early to apply defaults anyway – seeing a value in the config
+        // doesn't mean it's "active"
+        config.extend(Self::default().0.into_iter().filter(|_| true));
+
+        Ok(Self(config))
+    }
+
+    pub fn for_host(_: &str) -> Self {
         unimplemented!()
     }
 
-    pub fn with_config_file(self: &Self, _: &str) -> SSHConfig {
+    pub fn with_config_file(self: &Self, _: &str) -> Self {
         unimplemented!()
     }
 
     pub fn connect_with_auth(self: &Self) -> ssh2::Session {
         unimplemented!()
     }
+
+    fn user_dir() -> PathBuf {
+        [homedir(), OsString::from(".ssh")].iter().collect()
+    }
+
+    fn system_dir() -> PathBuf {
+        ["/etc", "ssh"].iter().collect()
+    }
+
+    pub fn user_config() -> PathBuf {
+        Self::user_dir().join("config")
+    }
+
+    pub fn system_config() -> PathBuf {
+        Self::system_dir().join("ssh_config")
+    }
+}
+
+// Similar to SSH's `flags` parameter
+#[derive(Copy, Clone)]
+struct ReadMeta {
+    depth: usize,
+    user_config: bool,
+}
+
+fn readconf_depth<P: AsRef<Path>>(path: P, meta: ReadMeta) -> Result<Vec<SSHOption>, Error> {
+    if meta.depth > MAX_READCONF_DEPTH {
+        return Err(Error::MaxDepthExceeded);
+    }
+
+    // The only depth 0 file that gets checked for perms is the user config file
+    if meta.depth > 0 || *path.as_ref() == SSHConfig::user_config() {
+        let meta = fs::metadata(&path)?;
+        let perms = meta.permissions();
+
+        if cfg!(unix) {
+            use std::os::unix::fs::MetadataExt;
+            use std::os::unix::fs::PermissionsExt;
+            if (meta.uid() != 0
+                && if cfg!(feature = "with_libc") {
+                    // SAFETY: getuid can never fail
+                    meta.uid() != unsafe { getuid() }
+                } else {
+                    false
+                })
+                || perms.mode() & 0o022 != 0
+            {
+                return Err(Error::PermissionError);
+            }
+        }
+    }
+
+    use option::Include::*;
+    use option::SSHOption::Include;
+    let file = File::open(path)?;
+    let mut opts = vec![];
+    for line in io::BufReader::new(file).lines() {
+        let opt = if let Some(opt) = option::parse_opt(line?)? {
+            opt
+        } else {
+            continue;
+        };
+
+        match opt {
+            Include(Paths(paths)) => {
+                let mut globbed = vec![];
+                for path in paths {
+                    /*
+                     * Ensure all paths are anchored. User configuration
+                     * files may begin with '~/' but system configurations
+                     * must not. If the path is relative, then treat it
+                     * as living in ~/.ssh for user configurations or
+                     * /etc/ssh for system ones.
+                     */
+                    let p = match path {
```

Comment:
It would be nice if p had type PathBuf. That would require `glob` to work in terms of paths, though, and not `&str`.

Reply:
Ended up going the other way, and rejecting all non-Unicode files and paths.